File: 백준/Gold/10423. 전기가 부족해/전기가 부족해.cc
// Authored by : chjh2129
#include <bits/stdc++.h>
/*
  발전소가 설치된 도시들은 서로 떨어져있지만 케이블로 연결될 필요도 없습니다.

  따라서 가상의 정점 0번을 두고 발전소가 설치된 도시들을 0번 도시가 가중치가 0인 케이블이 있다고 가정합니다.

  이를 인접리스트에 추가한다음 n+1개의 정점을 가진 그래프에서 최소 신장 트리를 구하면 됩니다.
*/
using namespace std;
using ti = tuple<int, int, int>;
using pi = pair<int, int>;

int n, m, k;
vector<pi> adj[1005];
vector<bool> chk(1005);
priority_queue<ti, vector<ti>, greater<ti>> pq;

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> n >> m >> k;
  
  while(k--){
    int x; cin >> x;
    adj[0].push_back({0, x});
    adj[x].push_back({0, 0});
  }

  while(m--){
    int u, v, cost;
    cin >> u >> v >> cost;
    adj[u].push_back({cost, v});
    adj[v].push_back({cost, u});
  }

  // solve
  chk[0] = 1;
  for(auto &[cost, nxt] : adj[0]) pq.push({cost, 0, nxt});

  int cnt{}, ans{};
  while(cnt < n){
    auto [cost, prv, cur] = pq.top(); pq.pop();
    if(chk[cur]) continue;
    chk[cur] = 1;
    ans += cost;
    cnt++;
    for(auto &[cost, nxt] : adj[cur]){
      if(chk[nxt]) continue;
      pq.push({cost, cur, nxt});
    }
  }

  // output
  cout << ans;
}

File: 백준/Gold/10423. 전기가 부족해/전기가 부족해.py
#Authored by : chjh2129
import sys
'''
  발전소가 설치된 도시들은 서로 떨어져있지만 케이블로 연결될 필요도 없습니다.

  따라서 가상의 정점 N+1번을 두고 발전소가 설치된 도시는 미리 연결을 해놓습니다.

  정점의 개수는 N+1이 되고 간선은 총 N개가 필요합니다.
'''
input = sys.stdin.readline

def find(x: int, p: list[int]) -> int:
  root: int = x
  while p[root] > 0: root = p[root]
  while p[x] > 0:
    nxt, p[x] = p[x], root
    x = nxt
  return root

def union(x: int, y: int, p: list[int]) -> bool:
  x = find(x, p)
  y = find(y, p)
  if x == y: return False
  if p[x] == p[y]: p[x] -= 1
  if p[x] > p[y]: x, y = y, x
  p[y] = x
  return True

def main() -> None:
  # input & init
  N, M, K = map(int, input().split())

  cnt: int = 0
  P: list[int] = [ -1 for _ in range(N+2) ]
  for cur in list(map(int, input().split())): 
    union(N+1, cur, P)
    cnt += 1
  
  Edge: list[tuple] = [ (cost, u, v) for u, v, cost in (map(int, input().split()) for _ in range(M)) ]

  # solve
  Edge.sort(key = lambda x: x[0])

  ans:int = 0
  for cost, u, v in Edge:
    if cnt == N: break
    if union(u, v, P):
      cnt += 1
      ans += cost
  
  # output
  print(ans)


sys.exit(main())